### ISSUE TITLE
[DataGrid] core v4 and dialog creates an infinite loop

### ISSUE THREAD

**Issue by Kogeer** [mui-org/material-ui-x#1464]
Dear Mui,

I found an issue. I hope it is correct and help to resolve an unexpected behavior.
I created a create-react-app. The App.js code is below:
```jsx
import {useState} from "react";
import {DataGrid} from "@material-ui/data-grid";
import {Dialog, IconButton} from "@material-ui/core";
import AddToPhotosIcon from '@material-ui/icons/AddToPhotos';

const App = () => {
  const [open, setOpen] = useState(false);
  const handleOpen = () => {
    setOpen(true);
  }
  const handleClose = () => {
    setOpen(false);
  }
  const rows = [
    {
      id: 1,
      name: 'Group1'
    },
    {
      id: 2,
      name: 'Group2'
    },
    {
      id: 3,
      name: 'Group3'
    }
  ];
  const columns = [
    {
      field: 'name',
      headerName: 'Group name',
      flex: 1,
    },
    {
      field: 'actions',
      headerName: 'Actions',
      renderCell: () => {
        return (
            <IconButton onClick={() => handleOpen()}>
              <AddToPhotosIcon />
            </IconButton>
        )
      }
    }
  ]

  return (
      <div style={{height: '300px'}}>
        <DataGrid
            columns={columns}
            rows={rows}
            onCellBlur={() => {
              console.log('on cell blur');
            }}
        />
        <Dialog open={open} onClose={handleClose} >
          {'HOLA'}
        </Dialog>
      </div>
  )
}

export default App;
```

https://codesandbox.io/s/material-demo-forked-j160c?file=/demo.tsx

- [x] The issue is present in the latest release.
- [ ] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Whit these dependencies it works fine:
"@material-ui/core": "^4.11.3",
"@material-ui/data-grid": "^4.0.0-alpha.26",

But i upgraded to:
 "@emotion/react": "^11.1.5",
 "@emotion/styled": "^11.3.0",
 "@material-ui/core": "^5.0.0-alpha.28",
 "@material-ui/data-grid": "^4.0.0-alpha.26",

After upgrade i have a current behavior which is stuck in onCellBlur loop.

## Expected Behavior 🤔

It works as early version. Just open the Dialog normally.

## Steps to Reproduce 🕹

Steps:

1. Create-react-app
2. Install upgraded versions (detaild in current behavior)
3. First click is in Actions cell next to button (not on AddToPhotos icon just in cell)
4. Second click on AddToPhotos icon which open the dialog.
5. onCellBlur event stucked picture here: https://imgur.com/DizToq5

## Context 🔦

I just want to use a button which open a Dialog.

## Your Environment 🌎

<details>
  <summary>`npx @material-ui/envinfo`</summary>
  
```
  System:
    OS: macOS 11.2.3
  Binaries:
    Node: 14.15.1 - ~/.nvm/versions/node/v14.15.1/bin/node
    Yarn: Not Found
    npm: 6.14.8 - ~/.nvm/versions/node/v14.15.1/bin/npm
  Browsers:
    Chrome: 90.0.4430.85
    Edge: Not Found
    Firefox: Not Found
    Safari: 14.0.3
  npmPackages:
    @emotion/react: ^11.1.5 => 11.1.5
    @emotion/styled: ^11.3.0 => 11.3.0
    @material-ui/core: ^5.0.0-alpha.28 => 5.0.0-alpha.31
    @material-ui/data-grid: ^4.0.0-alpha.26 => 4.0.0-alpha.26
    @material-ui/icons: ^4.11.2 => 4.11.2
    @material-ui/styled-engine:  5.0.0-alpha.26
    @material-ui/styles:  5.0.0-alpha.31
    @material-ui/system:  5.0.0-alpha.30
    @material-ui/types:  5.1.8
    @material-ui/unstyled:  5.0.0-alpha.31
    @material-ui/utils:  5.0.0-alpha.30
    @types/react:  17.0.3
    react: ^17.0.2 => 17.0.2
    react-dom: ^17.0.2 => 17.0.2
```
</details>
I'm using Chrome.

**Comment by dtassone** [comment]
Try to wrap your onCellBlur handler with `React.useCallback`.
You're setting a new handler at each render that might create issues.

**Comment by Kogeer** [comment]
Dear @dtassone,

I don't want to use any handler on onCellBlur, i just put `console.log()` for see where is the loop.
I forked a code-sandbox whit new dependencies what i wrote (which made @oliviertassinari and put in my issue). I updated the dependencies. Link here: https://codesandbox.io/s/material-demo-forked-cd2bb?file=/demo.tsx

You think try to wrap the `handleOpen()` handler in useCallBack? Because i tried, and the result is same unfortunately. If i delete the onCellBlur handling on DataGrid, the issue is still exist.

**Comment by Kogeer** [comment]
Sorry maybe i'm not clear and correct.

I just see `[as setCellFocus]` in error boundaries and just try to find which action maybe use this, that is why i put `console.log()` in many onCell... DataGrid actions. And somewhy in onCellBlur action just called many times. 
And with 4.11.3 @material-ui/core it is works nicely just open the dialog with the code above. But with the latest (5.0.xx) it is in loop.

**Comment by dtassone** [comment]
Got it!

To reproduce
1. Click action header to sort
2. Click on any action cell

Works on @material-ui/core 4.11.3

**Comment by Kogeer** [comment]
Oh really, and you can reproduce if select a row within actions cell and after that click on action button.
Thanks!

**Comment by oliviertassinari** [comment]
The issue is fixed in 4.0.0-alpha.27. @dtassone has added a blur handler to allow the focus enforce logic to "let it go": #1421. So again, an issue with calling .focus() in an effect.

**Comment by Kogeer** [comment]
Yes, perfect, very nice guys! :)
